File: targets/emojipaint/epilogue.js
import { model_init } from "./prelude";
import { update } from "./sketch";
// EMOJIPAINT MVU EPILOGUE
const test1 = () => {
    const [grid, ,] = update(model_init, { type: "StampEmoji", row: 0, col: 0 });
    return {
        result: grid[0][0] === "😄" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "",
        values: [grid, [["😄", "", ""], ["", "", ""], ["", "", ""]]],
    };
};
const test2 = () => {
    const [grid, ,] = update(model_init, { type: "FillRow", row: 1 });
    return {
        result: grid[0][0] === "" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "😄" && grid[1][1] === "😄" && grid[1][2] === "😄" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "",
        values: [grid, [["", "", ""], ["😄", "😄", "😄"], ["", "", ""]]],
    };
};
const test3 = () => {
    let model = update(model_init, { type: "SelectEmoji", emoji: "😅" });
    const [grid, selectedEmoji,] = update(model, { type: "StampEmoji", row: 2, col: 2 });
    return {
        result: grid[0][0] === "" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "😅" && selectedEmoji === "😅",
        values: [grid, [["", "", ""], ["", "", ""], ["", "", "😅"]], selectedEmoji, "😅"],
    };
};
const test4 = () => {
    let model = update(model_init, { type: "FillRow", row: 0 });
    const [grid, ,] = update(model, { type: "ClearCell", row: 0, col: 1 });
    return {
        result: grid[0][0] === "😄" && grid[0][1] === "" && grid[0][2] === "😄" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "",
        values: [grid, [["😄", "", "😄"], ["", "", ""], ["", "", ""]]],
    };
};
const test5 = () => {
    let model = update(model_init, { type: "StampEmoji", row: 1, col: 1 });
    const [grid, ,] = update(model, { type: "ClearGrid" });
    return {
        result: grid[0][0] === "" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "",
        values: [grid, [["", "", ""], ["", "", ""], ["", "", ""]]],
    };
};
const test6 = () => {
    const [, selectedEmoji,] = update(model_init, { type: "SelectEmoji", emoji: "😊" });
    const [grid_init, , emojiList_init] = model_init;
    const [grid, ,] = update([grid_init, selectedEmoji, emojiList_init], { type: "StampEmoji", row: 1, col: 2 });
    return {
        result: grid[0][0] === "" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "😊" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "",
        values: [grid, [["", "", ""], ["", "", "😊"], ["", "", ""]]],
    };
};
const test7 = () => {
    const [, selectedEmoji, emojiList] = model_init;
    const model = update(model_init, { type: "FillRow", row: 2 });
    const [grid, ,] = update(model, { type: "ClearCell", row: 2, col: 0 });
    return {
        result: grid[0][0] === "" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "" && grid[2][0] === "" && grid[2][1] === "😄" && grid[2][2] === "😄",
        values: [grid, [["", "", ""], ["", "", ""], ["", "😄", "😄"]]],
    };
};
const test8 = () => {
    let model = update(model_init, { type: "StampEmoji", row: 0, col: 0 });
    model = update(model, { type: "StampEmoji", row: 1, col: 1 });
    model = update(model, { type: "StampEmoji", row: 2, col: 2 });
    const [grid, ,] = update(model, { type: "ClearGrid" });
    return {
        result: grid[0][0] === "" && grid[0][1] === "" && grid[0][2] === "" && grid[1][0] === "" && grid[1][1] === "" && grid[1][2] === "" && grid[2][0] === "" && grid[2][1] === "" && grid[2][2] === "",
        values: [grid, [["", "", ""], ["", "", ""], ["", "", ""]]],
    };
};
const test9 = () => {
    const [grid_init, , emojiList_init] = model_init;
    let model = update(model_init, { type: "FillRow", row: 0 });
    const [, selectedEmoji,] = update(model, { type: "SelectEmoji", emoji: "😆" });
    const [grid, ,] = model;
    const [updatedGrid, ,] = update([grid, selectedEmoji, emojiList_init], { type: "StampEmoji", row: 1, col: 1 });
    return {
        result: updatedGrid[0][0] === "😄" && updatedGrid[0][1] === "😄" && updatedGrid[0][2] === "😄" && updatedGrid[1][0] === "" && updatedGrid[1][1] === "😆" && updatedGrid[1][2] === "" && updatedGrid[2][0] === "" && updatedGrid[2][1] === "" && updatedGrid[2][2] === "",
        values: [updatedGrid, [["😄", "😄", "😄"], ["", "😆", ""], ["", "", ""]]],
    };
};
const test10 = () => {
    let model = update(model_init, { type: "StampEmoji", row: 0, col: 0 });
    model = update(model, { type: "FillRow", row: 2 });
    const [grid, , emojiList] = model;
    model = update(model, { type: "SelectEmoji", emoji: "😉" });
    model = update(model, { type: "StampEmoji", row: 1, col: 1 });
    model = update(model, { type: "ClearCell", row: 2, col: 2 });
    const [updatedGrid, selectedEmoji, _] = model;
    return {
        result: updatedGrid[0][0] === "😄" && updatedGrid[0][1] === "" && updatedGrid[0][2] === "" && updatedGrid[1][0] === "" && updatedGrid[1][1] === "😉" && updatedGrid[1][2] === "" && updatedGrid[2][0] === "😄" && updatedGrid[2][1] === "😄" && updatedGrid[2][2] === "" && selectedEmoji === "😉",
        values: [updatedGrid, [["😄", "", ""], ["", "😉", ""], ["😄", "😄", ""]], selectedEmoji, "😉"],
    };
};
const tests = [test1, test2, test3, test4, test5, test6, test7, test8, test9, test10];
let score = 0;
for (let i = 0; i < tests.length; ++i) {
    try {
        const run = tests[i]();
        console.assert(run.result === true, "%o", { i: i + 1, values: run.values });
        if (run.result) {
            score++;
        }
    }
    catch (err) {
        console.log(err);
    }
}
console.log(`score: ${score} / ${tests.length}`);

File: targets/ocaml/emojipaint/epilogue.ml
(* EMOJIPAINT MVU TESTS *)
open Prelude
open Sketch
open OUnit2

(* test  *)
(*   let (grid, _, _) = update(model_init, StampEmoji(0, 0)) in *)
(*   grid == [["😄","",""],["","",""],["","",""]]  *)
(* end; *)
let test1 test_ctxt =
  let grid, _, _ = update (model_init, StampEmoji (0, 0)) in
  assert_equal grid [ [ "😄"; ""; "" ]; [ ""; ""; "" ]; [ ""; ""; "" ] ]

(* test *)
(*   let (grid, _, _) = update(model_init, FillRow(1)) in *)
(*   grid == [["","",""],["😄","😄","😄"],["","",""]]  *)
(* end; *)
let test2 test_ctxt =
  let grid, _, _ = update (model_init, FillRow 1) in
  assert_equal grid [ [ ""; ""; "" ]; [ "😄"; "😄"; "😄" ]; [ ""; ""; "" ] ]

(* test *)
(*   let model = update(model_init, SelectEmoji("😅")) in *)
(*   let (grid, selectedEmoji, _) = update(model, StampEmoji(2, 2)) in *)
(*   grid == [["","",""],["","",""],["","","😅"]] && selectedEmoji == "😅"  *)
(* end; *)
let test3 test_ctxt =
  let model = update (model_init, SelectEmoji "😅") in
  let grid, selectedEmoji, _ = update (model, StampEmoji (2, 2)) in
  assert_bool "test3 failed"
    (grid = [ [ ""; ""; "" ]; [ ""; ""; "" ]; [ ""; ""; "😅" ] ]
    && selectedEmoji = "😅")

(* test *)
(*   let model = update(model_init, FillRow(0)) in *)
(*   let (grid, _, _) = update(model, ClearCell(0, 1)) in *)
(*   grid == [["😄","","😄"],["","",""],["","",""]]  *)
(* end; *)
let test4 test_ctxt =
  let model = update (model_init, FillRow 0) in
  let grid, _, _ = update (model, ClearCell (0, 1)) in
  assert_equal grid [ [ "😄"; ""; "😄" ]; [ ""; ""; "" ]; [ ""; ""; "" ] ]

(* test *)
(*   let model = update(model_init, StampEmoji(1, 1)) in *)
(*   let (grid, _, _) = update(model, ClearGrid) in *)
(*   grid == [["","",""],["","",""],["","",""]]  *)
(* end; *)
let test5 test_ctxt =
  let model = update (model_init, StampEmoji (1, 1)) in
  let grid, _, _ = update (model, ClearGrid) in
  assert_equal grid [ [ ""; ""; "" ]; [ ""; ""; "" ]; [ ""; ""; "" ] ]

(* test *)
(*   let (_, selectedEmoji, _) = update(model_init, SelectEmoji("😊")) in *)
(*   let (grid_init, _, emojiList_init) = model_init in *)
(*   let (grid, _, _) = update((grid_init, selectedEmoji, emojiList_init), StampEmoji(1, 2)) in *)
(*   grid == [["","",""],["","","😊"],["","",""]]  *)
(* end; *)
let test6 test_ctxt =
  let _, selectedEmoji, _ = update (model_init, SelectEmoji "😊") in
  let grid_init, _, emojiList_init = model_init in
  let grid, _, _ =
    update ((grid_init, selectedEmoji, emojiList_init), StampEmoji (1, 2))
  in
  assert_equal grid [ [ ""; ""; "" ]; [ ""; ""; "😊" ]; [ ""; ""; "" ] ]

(* test *)
(*   let (_, selectedEmoji, emojiList) = model_init in *)
(*   let model = update(model_init, FillRow(2)) in *)
(*   let (grid, _, _) = update(model, ClearCell(2, 0)) in *)
(*   grid == [["","",""],["","",""],["","😄","😄"]]  *)
(* end; *)
let test7 test_ctxt =
  let _, selectedEmoji, emojiList = model_init in
  let model = update (model_init, FillRow 2) in
  let grid, _, _ = update (model, ClearCell (2, 0)) in
  assert_equal grid [ [ ""; ""; "" ]; [ ""; ""; "" ]; [ ""; "😄"; "😄" ] ]

(* test *)
(*   let model = update(model_init, StampEmoji(0, 0)) in *)
(*   let model = update(model, StampEmoji(1, 1)) in *)
(*   let model = update(model, StampEmoji(2, 2)) in *)
(*   let (grid, _, _) = update(model, ClearGrid) in *)
(*   grid == [["","",""],["","",""],["","",""]]  *)
(* end; *)
let test8 test_ctxt =
  let model = update (model_init, StampEmoji (0, 0)) in
  let model = update (model, StampEmoji (1, 1)) in
  let model = update (model, StampEmoji (2, 2)) in
  let grid, _, _ = update (model, ClearGrid) in
  assert_equal grid [ [ ""; ""; "" ]; [ ""; ""; "" ]; [ ""; ""; "" ] ]

(* test *)
(*   let (grid_init, _, emojiList_init) = model_init in *)
(*   let model = update(model_init, FillRow(0)) in *)
(*   let (_, selectedEmoji, _) = update(model, SelectEmoji("😆")) in *)
(*   let (grid,_,_) = model in *)
(*   let (grid, _, _) = update((grid, selectedEmoji, emojiList_init), StampEmoji(1, 1)) in *)
(*   grid == [["😄","😄","😄"],["","😆",""],["","",""]]  *)
(* end; *)
let test9 test_ctxt =
  let grid_init, _, emojiList_init = model_init in
  let model = update (model_init, FillRow 0) in
  let _, selectedEmoji, _ = update (model, SelectEmoji "😆") in
  let grid, _, _ = model in
  let grid, _, _ =
    update ((grid, selectedEmoji, emojiList_init), StampEmoji (1, 1))
  in
  assert_equal grid [ [ "😄"; "😄"; "😄" ]; [ ""; "😆"; "" ]; [ ""; ""; "" ] ]

(* test *)
(*   let model = update(model_init, StampEmoji(0, 0)) in *)
(*   let model = update(model, FillRow(2)) in *)
(*   let (grid, _, emojiList) = model in *)
(*   let model = update(model, SelectEmoji("😉")) in *)
(*   let model = update(model, StampEmoji(1, 1)) in *)
(*   let model = update(model, ClearCell(2, 2)) in *)
(*   let (grid, selectedEmoji, _) = model in *)
(*   grid == [["😄","",""],["","😉",""],["😄","😄",""]] && selectedEmoji == "😉"  *)
(* end; *)
let test10 test_ctxt =
  let model = update (model_init, StampEmoji (0, 0)) in
  let model = update (model, FillRow 2) in
  let grid, _, emojiList = model in
  let model = update (model, SelectEmoji "😉") in
  let model = update (model, StampEmoji (1, 1)) in
  let model = update (model, ClearCell (2, 2)) in
  let grid, selectedEmoji, _ = model in
  assert_bool "test10 failed"
    (grid = [ [ "😄"; ""; "" ]; [ ""; "😉"; "" ]; [ "😄"; "😄"; "" ] ]
    && selectedEmoji = "😉")
